apps/system/js/screen_auto_brightness.js

```diff
@@ -0,0 +1,136 @@
+'use strict';
+
+(function(exports) {
+
+  /*
+   * constructor - argument is the function that will adjust
+   * the screen brightness. This function takes 2 arguments,
+   * the brightness and whether the change is immediate (true),
+   * or will be transitioned (false)
+   */
+  var ScreenAutoBrightness = function ScreenAutoBrightness(
+      setScreenBrightnessFn) {
+    this._setScreenBrightness = setScreenBrightnessFn;
+  };
+
+  /*
+   * The auto-brightness algorithm will never set the screen brightness
+   * to a value smaller than this. 0.1 seems like a good screen brightness
+   * in a completely dark room on a Unagi.
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_MINIMUM = 0.1;
+
+  /*
+   * This constant is used in the auto brightness algorithm. We take
+   * the base 10 logarithm of the incoming lux value from the light
+   * sensor and multiplied it by this constant. That value is used to
+   * compute a weighted average with the current brightness and
+   * finally that average brightess is and then clamped to the range
+   * [AUTO_BRIGHTNESS_MINIMUM, 1.0].
+   *
+   * Making this value larger will increase the brightness for a given
+   * ambient light level. At a value of about .25, the screen will be
+   * at full brightness in sunlight or in a well-lighted work area.
+   * At a value of about .3, the screen will typically be at maximum
+   * brightness in outdoor daylight conditions, even when overcast.
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_CONSTANT = 0.27;
+
+  /*
+   * We won't set a new brightness value if the difference between the old and
+   * new ambient light sensor values is lower than this constant.
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_MIN_DELTA = 10;
+
+  /*
+   * This variable contains the latest ambient light sensor value that was used
+   * to set a brightness.
+   */
+  ScreenAutoBrightness.prototype._previousLux = undefined;
+
+  /*
+   * auto adjust screen delays, see autoAdjustBrightness function
+   * wait for _autoAdjustDelay milliseconds before adjusting brightness value
+   */
+  ScreenAutoBrightness.prototype._autoAdjustState = 0;
+  ScreenAutoBrightness.prototype._autoAdjustDelay = 750;
+
+  /*
+   * keep track of last lux value during autoAdjustBrightness delay's
+   */
+  ScreenAutoBrightness.prototype._autoDelayPrevLux = undefined;
+
+  // reinitialize to start state
+  ScreenAutoBrightness.prototype.init = function() {
```

Comment:
This should be `reset()`. You could provide a `start` function that simply do `this.reset()` as well.

The convention is having `start()` and `stop()` being called *once* when the instance is created/destroyed. Since you intend to have ScreenManager to reuse the instance already created, let's not use `start()` and use `reset()`.

`init()` is everywhere but unfortunately it's not well defined in our old code pattern so let's not copy it.